File: examples/load-persona.mjs
#!/usr/bin/env node
/**
 * Load a persona YAML by id and build a chat-completions messages array.
 *
 *   node examples/load-persona.mjs ada-vale "The p99 doubled after retries."
 */

import { readFileSync, readdirSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import { parse } from "yaml";

const root = join(dirname(fileURLToPath(import.meta.url)), "..");

export function findPersonaPath(id) {
  const categories = [
    "professional",
    "creative",
    "personality",
    "specialized",
    "experimental",
    "business",
  ];
  for (const category of categories) {
    const path = join(root, "personas", category, `${id}.yaml`);
    try {
      readFileSync(path);
      return path;
    } catch {
      /* try next */
    }
  }
  const known = categories.flatMap((category) =>
    readdirSync(join(root, "personas", category))
      .filter((n) => n.endsWith(".yaml"))
      .map((n) => n.replace(/\.yaml$/, "")),
  );
  throw new Error(`Unknown persona "${id}". Known: ${known.join(", ")}`);
}

export function loadPersona(id) {
  return parse(readFileSync(findPersonaPath(id), "utf8"));
}

export function toMessages(persona, userText, options = {}) {
  const messages = [{ role: "system", content: persona.system_prompt }];
  if (options.withExamples) {
    const n = Number(options.withExamples) || 1;
    for (const ex of persona.example_interactions.slice(0, n)) {
      messages.push({ role: "user", content: ex.user });
      messages.push({ role: "assistant", content: ex.assistant });
    }
  }
  messages.push({ role: "user", content: userText });
  return messages;
}

const isMain = process.argv[1] && fileURLToPath(import.meta.url) === process.argv[1];
if (isMain) {
  const id = process.argv[2];
  const prompt = process.argv.slice(3).join(" ") || "Introduce yourself in four sentences.";
  if (!id) {
    console.error("usage: node examples/load-persona.mjs <id> [user prompt]");
    process.exit(2);
  }
  const persona = loadPersona(id);
  const messages = toMessages(persona, prompt);
  console.log(JSON.stringify({ id: persona.id, name: persona.name, messages }, null, 2));
}

File: examples/load_persona.py
#!/usr/bin/env python3
"""Load a persona YAML by id and build a chat-completions messages array.

    python examples/load_persona.py ada-vale "The p99 doubled after retries."
"""

from __future__ import annotations

import json
import sys
from pathlib import Path

try:
    import yaml
except ImportError as exc:  # pragma: no cover
    raise SystemExit("Install PyYAML: pip install pyyaml") from exc

ROOT = Path(__file__).resolve().parents[1]
CATEGORIES = (
    "professional",
    "creative",
    "personality",
    "specialized",
    "experimental",
    "business",
)


def find_persona_path(persona_id: str) -> Path:
    for category in CATEGORIES:
        path = ROOT / "personas" / category / f"{persona_id}.yaml"
        if path.exists():
            return path
    known = [
        p.stem
        for category in CATEGORIES
        for p in (ROOT / "personas" / category).glob("*.yaml")
    ]
    raise FileNotFoundError(f'Unknown persona "{persona_id}". Known: {", ".join(known)}')


def load_persona(persona_id: str) -> dict:
    with find_persona_path(persona_id).open(encoding="utf-8") as handle:
        return yaml.safe_load(handle)


def to_messages(persona: dict, user_text: str, with_examples: int = 0) -> list[dict]:
    messages = [{"role": "system", "content": persona["system_prompt"]}]
    if with_examples:
        for example in persona["example_interactions"][:with_examples]:
            messages.append({"role": "user", "content": example["user"]})
            messages.append({"role": "assistant", "content": example["assistant"]})
    messages.append({"role": "user", "content": user_text})
    return messages


def main(argv: list[str]) -> int:
    if len(argv) < 2:
        print("usage: python examples/load_persona.py <id> [user prompt]", file=sys.stderr)
        return 2
    persona_id = argv[1]
    prompt = " ".join(argv[2:]) or "Introduce yourself in four sentences."
    persona = load_persona(persona_id)
    payload = {
        "id": persona["id"],
        "name": persona["name"],
        "messages": to_messages(persona, prompt),
    }
    print(json.dumps(payload, indent=2))
    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv))
